src/librustdoc/doctest.rs

```diff
@@ -249,14 +265,82 @@ pub(crate) fn run(
 pub(crate) fn run_tests(
     mut test_args: Vec<String>,
     nocapture: bool,
-    mut tests: Vec<test::TestDescAndFn>,
+    opts: GlobalTestOptions,
+    rustdoc_options: &Arc<RustdocOptions>,
+    unused_extern_reports: &Arc<Mutex<Vec<UnusedExterns>>>,
+    mut standalone_tests: Vec<test::TestDescAndFn>,
+    mergeable_tests: FxHashMap<Edition, Vec<(DocTest, ScrapedDoctest)>>,
```

Comment:
Instead of grouping solely on `Edition`, I think it would be better to key by a new datatype `DoctestGroup` that would be defined like this:

```rust
struct DoctestGroup {
  edition: Edition,
  stdlib: IncludedStdlib,
  test_harness: bool,
  nocapture: bool, // maybe?
}

enum IncludedStdlib {
  All,
  NoStd,
  NoCore,
}
```

We can then group doctests more aggressively, which could be especially valuable in crates that use `#![no_std]` a lot in their doctests.

Reply:
I'm not quite sure what you mean. My comment has nothing to do with somehow supporting tests that use both `#![no_std]` and `test_harness`. I'm just suggesting a way we can merge more doctests than the current implementation does.

The current implementation in this PR looks at each test together with the args passed to rustdoc and decides which tests are mergeable and which must be compiled and run on their own. Then it merges all mergeable tests into one file for each edition, since obviously doctests of different editions can't be in the same rustc invocation. It considers tests that use the langstring `test_harness` (which allows using `#[test]` in a doctest) or the attribute `#![no_std]` to be unmergeable (standalone). I'm suggesting that we instead allow merging these tests, but only with other tests that match this attribute.

Please let me know if I misunderstood your comment.

Reply:
Hmm, thank you for clarifying. I'll take another look at the impl.